Challenge sum of digits

SEMIRATESFAI opened the pull request:
---
name: solution review 
about: A template PR for code review with a checklist
---

<!--
  make this PR easy to find:

  - assign: yourself
  - link it to the corresponding challenge issue on the project board
-->

<!-- describe the challenge you will solve -->

## Behavior: Write a function that takes a positive integer and returns the sum of its digits. 

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

## Strategy

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

SEMIRATESFAI committed:
create the folders and the init files

SEMIRATESFAI committed:
Merge branch 'main' into challenge-sum-of-digits

SEMIRATESFAI committed:
add the solution to the sum_of_digits_challenge and the test

SEMIRATESFAI committed:
Update sum_of_digits.py

Refined sum_of_digits code and tests: added docstrings, improved test cases, and ensured all linting checks passed.

SEMIRATESFAI committed:
Update test_sum_of_digits.py

Refined sum_of_digits code and tests: added docstrings, improved test cases, and ensured all linting checks passed.

SEMIRATESFAI committed:
made changes to the code by adding headings to the steps

MPKenley commented on solutions/challenge_9/sum_of_digits.py:
```diff
@@ -0,0 +1,42 @@
+    >>> sum_of_digits(0)
+    0
+    """
+    assert isinstance(n, int) and n >= 0, "input must be a positive integer"
```
Inputs Validation: assert in the module of the challenge


Feedback : Think about replacing the _assert_ statement used for input validation by an explicite if statement and a valueError. Because _assert_ is not optimize for input validation. It can be disabled in certains environnement

SEMIRATESFAI replied:
Yes, changes has been done, making these updates will ensure the code robustly handles invalid inputs and meets the feedback requirements. Thanks for your feedback!

MPKenley commented on solutions/tests/challenge_9/test_sum_of_digits.py:
```diff
@@ -0,0 +1,38 @@
+class TestSumOfDigits(unittest.TestCase):
+    """Tests for the sum_of_digits function."""
+
+    def test_positive_integer(self):
```
Test names: 


Test names can be more descriptive to reflect the objective of the test. Like test_with_positive_integer instead of test_positive_integer.

SEMIRATESFAI replied:
Changes have been made. Thanks for your review.

MPKenley commented on solutions/tests/challenge_9/test_sum_of_digits.py:
```diff
@@ -0,0 +1,38 @@
+        """Ensure that the sum of digits of zero is zero."""
+        self.assertEqual(sum_of_digits(0), 0)
+
+    def test_invalid_input(self):
```
Test method logic: for test_invalid_input


feedback: separate logic within test methods into individual methods for clarity and simplicity.

SEMIRATESFAI replied:
Thank you so much for your thorough review and valuable feedback! I appreciate your efforts in helping me improve my code. 
Regarding the point about logic within the unit tests, upon review, it seems that the tests adhere to the best practices by focusing solely on verifying the expected behavior without including unnecessary logic. Each test method strictly checks the outputs and raises exceptions as intended.
I'd love to hear your thoughts on any specific changes you might suggest to further enhance the clarity and simplicity of the tests. Thanks again for your feedback!

MPKenley reviewed (changes requested):
1-input validation: Think about replacing the assert statement used for input validation by an explicite if statement and a valueError. Because assert is not optimize for input validation. It can be disabled in certains environnement

2- Test names: Test names can be more descriptive to reflect the objective of the test. Like test_with_positive_integer instead of test_positive_integer

3- Test method logic: separate logic within test methods into individual methods for clarity and simplicity
This review is just about making the `sum_of_digits.py` and its test more clear, simple, understandable and meet all the requirements.

SEMIRATESFAI committed:
Update test_sum_of_digits.py

adding "with" in the test names provides a bit more context about what inputs are being tested against, making the test scenarios clearer and more descriptive for people to read it

SEMIRATESFAI committed:
Update sum_of_digits.py

Replace the assert statement with an explicit if statement and raise a ValueError in the sum_of_digits function

SEMIRATESFAI reviewed (commented):
I have made the requested changes. Could you please review them? If you're satisfied, kindly submit. Thanks so much for your help!

MPKenley reviewed (changes requested):
Good job Semira
I appreciate your hard work and dedication. 

But it seems some of your ruff tests don't pass. Check again before merging.

SEMIRATESFAI reviewed (commented):
Your reviews and suggestions were spot-on! Thanks for your feedback!

SEMIRATESFAI committed:
replaced the assert statement used for input validation by an explicite if statement and a valueError

SEMIRATESFAI committed:
corrected suggested changes from reviewer

SEMIRATESFAI committed:
updated changes that were recomended by reviewer

likechrisss committed:
Merge branch 'main' into challenge-sum-of-digits

SEMIRATESFAI committed:
Merge branch 'main' into challenge-sum-of-digits

MPKenley reviewed (approved):
Félicitations for the great work Semira

SEMIRATESFAI commented:
Thanks for your review!
